File: langs/csharp/src/VMx/Commands/AsyncRelayCommand.cs
using System.Collections.Immutable;
using System.Reactive;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using System.Reactive.Subjects;

namespace VMx.Commands;

/// <summary>
/// Concrete cancellable async <see cref="IAsyncCommand"/> implementation.
///
/// Spec: spec/04-commands.md §10, ADR-0056.
/// - Task receives a <see cref="CancellationToken"/> linked to both the in-flight
///   <see cref="Cancel"/> source and any token supplied to <see cref="ExecuteAsync"/>.
/// - Predicate null → <see cref="CanExecute"/> returns true when idle.
/// - While an execution is in flight, <see cref="CanExecute"/> returns false (so the
///   command cannot double-run), and <see cref="CanExecuteChanged"/> fires when the
///   in-flight state flips on start and on completion.
/// - Cancellation is NON-THROWING by default (DIA-007 alignment): the awaited
///   <see cref="ExecuteAsync"/> completes normally on cancel. Opt into the throwing
///   mode via the builder's <c>ThrowOnCancel()</c>.
/// - A faulted task (non-cancellation) propagates to the awaiter of
///   <see cref="ExecuteAsync"/>; on the fire-and-forget <see cref="Execute"/> path —
///   which has no caller to propagate to — it is routed to <see cref="Errors"/>
///   instead of being swallowed (mirrors ConfirmationDecoratorCommand, ADR-0049).
/// - Builder is IMMUTABLE (BLD-001): every setter returns a NEW builder instance.
/// </summary>
public sealed class AsyncRelayCommand : IAsyncCommand, IDisposable
{
    private readonly Func<CancellationToken, Task> _task;
    private readonly Func<bool>? _predicate;
    private readonly bool _throwOnCancel;
    private readonly CompositeDisposable _triggerSubscriptions = new();
    private readonly Subject<Exception> _errors = new();
    private readonly object _gate = new();
    private CancellationTokenSource? _cts;
    private bool _isExecuting;
    private bool _disposed;
    // Set by Cancel()/Dispose() so the catch block swallows only a cancellation we
    // requested through the command's own channel; an externally-supplied token's
    // cancellation (flag unset) is re-raised per spec/04 §10.3.
    private volatile bool _cancelRequested;

    internal AsyncRelayCommand(
        Func<CancellationToken, Task>? task,
        Func<bool>? predicate,
        bool throwOnCancel,
        IReadOnlyList<IObservable<Unit>> triggers)
    {
        _task = task ?? (_ => Task.CompletedTask);
        _predicate = predicate;
        _throwOnCancel = throwOnCancel;
        foreach (var t in triggers)
            _triggerSubscriptions.Add(t.Subscribe(_ => RaiseCanExecuteChanged()));
    }

    /// <inheritdoc/>
    public event EventHandler? CanExecuteChanged;

    /// <inheritdoc/>
    public bool IsExecuting
    {
        get
        {
            lock (_gate)
            {
                return _isExecuting;
            }
        }
    }

    /// <summary>
    /// Surfaces a fault from the fire-and-forget <see cref="Execute"/> path (a throwing
    /// task that is not a cancellation). The awaitable <see cref="ExecuteAsync"/> path
    /// keeps its throw behavior — await it directly to handle the error inline.
    /// Cancellations never reach this channel. Completes on <see cref="Dispose"/>.
    /// </summary>
    public IObservable<Exception> Errors => _errors.AsObservable();

    /// <summary>
    /// Returns false while an execution is in flight; otherwise true if no predicate is
    /// configured, or the predicate's result. A throwing predicate returns false.
    /// </summary>
    public bool CanExecute(object? parameter)
    {
        lock (_gate)
        {
            if (_disposed || _isExecuting) return false;
        }
        if (_predicate is null) return true;
        try { return _predicate(); }
        catch { return false; }
    }

    /// <summary>
    /// Fire-and-forget entry-point: starts <see cref="ExecuteAsync"/> and returns
    /// immediately. A faulting task is routed to <see cref="Errors"/> (cancellation is
    /// already swallowed by the non-throwing default).
    /// </summary>
    public void Execute(object? parameter) =>
        _ = ExecuteAsync(parameter).ContinueWith(
            t => EmitError(t.Exception!.GetBaseException()),
            CancellationToken.None,
            TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously,
            TaskScheduler.Default);

    // Routes a fire-and-forget fault to the Errors channel. The disposed check
    // is taken under _gate — the continuation runs on TaskScheduler.Default,
    // concurrently with a possible Dispose() (whose _disposed write is gated).
    // Reading _disposed off-gate was a data race on a non-volatile field. The
    // try/catch closes the residual window where Dispose() completes+disposes
    // _errors between the gated check and the emit: System.Reactive's
    // Subject.OnNext throws ObjectDisposedException post-dispose, whereas the
    // other flavors' subjects no-op — dropping the fault here restores parity
    // rather than surfacing an unobserved task exception.
    private void EmitError(Exception error)
    {
        lock (_gate)
        {
            if (_disposed) return;
        }
        try
        {
            _errors.OnNext(error);
        }
        catch (ObjectDisposedException)
        {
            // Concurrent Dispose() tore down the error subject; drop the fault.
        }
    }

    /// <inheritdoc/>
    public async Task ExecuteAsync(object? parameter = null, CancellationToken cancellationToken = default)
    {
        var cts = TryBeginExecution(cancellationToken);
        if (cts is null) return;

        RaiseCanExecuteChanged();
        try
        {
            await _task(cts.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (!_throwOnCancel && _cancelRequested)
        {
            // Non-throwing cancellation (DIA-007 alignment): the cancel was requested
            // through this command's own channel (Cancel()/Dispose()), so complete
            // normally. An externally-supplied token's cancellation leaves
            // _cancelRequested false and is re-raised (spec/04 §10.3, parity with
            // Python/Swift).
        }
        finally
        {
            lock (_gate)
            {
                if (ReferenceEquals(_cts, cts))
                    _cts = null;
                _isExecuting = false;
            }
            cts.Dispose();
            RaiseCanExecuteChanged();
        }
    }

    /// <inheritdoc/>
    public void Cancel()
    {
        CancellationTokenSource? cts;
        lock (_gate)
        {
            _cancelRequested = true;
            cts = _cts;
        }
        cts?.Cancel();
    }

    /// <summary>
    /// Cancels any in-flight execution, disposes trigger subscriptions, and completes
    /// the <see cref="Errors"/> channel. Idempotent.
    /// </summary>
    public void Dispose()
    {
        CancellationTokenSource? cts;
        lock (_gate)
        {
            if (_disposed) return;
            _disposed = true;
            _cancelRequested = true;
            cts = _cts;
        }
        cts?.Cancel();
        _triggerSubscriptions.Dispose();
        _errors.OnCompleted();
        _errors.Dispose();
    }

    private CancellationTokenSource? TryBeginExecution(CancellationToken cancellationToken)
    {
        lock (_gate)
        {
            if (_disposed || _isExecuting) return null;
            if (_predicate is not null)
            {
                bool allowed;
                try { allowed = _predicate(); }
                catch { return null; }
                if (!allowed) return null;
            }

            var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            _cts = cts;
            _cancelRequested = false;
            _isExecuting = true;
            return cts;
        }
    }

    /// <summary>
    /// Notifies subscribers that <see cref="CanExecute"/> may have changed without
    /// evaluating the predicate or invoking the task. A no-op after disposal and
    /// valid both while idle and while an execution is in flight.
    /// </summary>
    public void RaiseCanExecuteChanged()
    {
        lock (_gate)
        {
            if (_disposed) return;
        }
        CanExecuteChanged?.Invoke(this, EventArgs.Empty);
    }

    /// <summary>Returns a new immutable builder for <see cref="AsyncRelayCommand"/>.</summary>
    public static AsyncRelayCommandBuilder Builder() => new();
}

/// <summary>
/// Immutable fluent builder for <see cref="AsyncRelayCommand"/> (BLD-001). Each setter
/// returns a NEW builder instance; the original is unchanged.
/// </summary>
public sealed class AsyncRelayCommandBuilder
{
    private readonly Func<CancellationToken, Task>? _task;
    private readonly Func<bool>? _predicate;
    private readonly bool _throwOnCancel;
    private readonly ImmutableList<IObservable<Unit>> _triggers;

    /// <summary>Creates an empty builder (no task, predicate, or triggers).</summary>
    public AsyncRelayCommandBuilder()
        : this(null, null, false, ImmutableList<IObservable<Unit>>.Empty)
    {
    }

    private AsyncRelayCommandBuilder(
        Func<CancellationToken, Task>? task,
        Func<bool>? predicate,
        bool throwOnCancel,
        ImmutableList<IObservable<Unit>> triggers)
    {
        _task = task;
        _predicate = predicate;
        _throwOnCancel = throwOnCancel;
        _triggers = triggers;
    }

    /// <summary>Sets the cancellable async task. Returns a new builder.</summary>
    public AsyncRelayCommandBuilder Task(Func<CancellationToken, Task> task)
        => new(task, _predicate, _throwOnCancel, _triggers);

    /// <summary>Sets the predicate gating <c>CanExecute</c>. Returns a new builder.</summary>
    public AsyncRelayCommandBuilder Predicate(Func<bool> predicate)
        => new(_task, predicate, _throwOnCancel, _triggers);

    /// <summary>Adds a trigger observable (additive). Returns a new builder.</summary>
    public AsyncRelayCommandBuilder Triggers(IObservable<Unit> trigger)
        => new(_task, _predicate, _throwOnCancel, _triggers.Add(trigger));

    /// <summary>
    /// Opts into the throwing-on-cancel mode (default off): when enabled, a cancelled
    /// execution surfaces the <see cref="OperationCanceledException"/> to the awaiter of
    /// <c>ExecuteAsync</c> rather than completing normally. Returns a new builder.
    /// </summary>
    public AsyncRelayCommandBuilder ThrowOnCancel(bool throwOnCancel = true)
        => new(_task, _predicate, throwOnCancel, _triggers);

    /// <summary>Builds the <see cref="AsyncRelayCommand"/> (succeeds even with no task).</summary>
    public AsyncRelayCommand Build()
        => new(_task, _predicate, _throwOnCancel, _triggers);
}

File: langs/python/src/vmx/commands/async_relay_command.py
"""AsyncRelayCommand — a cancellable async command.

Spec: spec/04-commands.md §10 (async command cancellation), ADR-0056.

Behavior contract:
- ``task`` is a coroutine function (``() -> Awaitable[None]``). asyncio delivers
  cancellation by raising ``CancelledError`` at the task's next await point — no
  explicit token is threaded through, matching idiomatic asyncio.
- ``can_execute`` returns False while an execution is in flight, so the command
  cannot double-run; it fires ``can_execute_changed`` when the in-flight state
  flips on start and on completion.
- ``cancel()`` cancels the in-flight task. Cancellation is NON-THROWING by default
  (alignment with the dialog cancellation contract, DIA-007 / spec ch.19 §7):
  the awaited ``execute_async`` returns normally on cancel. Opt into the throwing
  mode with ``throw_on_cancel=True``.
- A faulting task (non-cancellation) propagates to the awaiter of
  ``execute_async``; on the fire-and-forget ``execute()`` path — which has no
  caller to propagate to — it is routed to the :attr:`errors` channel instead of
  being swallowed (mirrors ``ConfirmationDecoratorCommand``, VMX-009 / ADR-0049).
- Builder is IMMUTABLE (BLD-001): every setter returns a NEW builder instance.
"""

from __future__ import annotations

import asyncio
import dataclasses
from collections.abc import Awaitable, Callable
from concurrent.futures import Future
from threading import RLock

import reactivex as rx
from reactivex import operators as ops
from reactivex.subject import Subject

from vmx._asyncio_runner import submit_background


class AsyncRelayCommand:
    """Cancellable async command built via an immutable fluent builder.

    Use ``AsyncRelayCommand.builder()`` to start a build.
    """

    def __init__(
        self,
        task: Callable[[], Awaitable[None]] | None,
        predicate: Callable[[], bool] | None,
        triggers: list[rx.Observable[object]],
        *,
        throw_on_cancel: bool = False,
    ) -> None:
        self._task = task
        self._predicate = predicate
        self._throw_on_cancel = throw_on_cancel
        self._gate = RLock()
        self._can_execute_changed_subject: Subject[None] = Subject()
        self._errors: Subject[BaseException] = Subject()
        self._current_task: asyncio.Task[None] | None = None
        self._current_loop: asyncio.AbstractEventLoop | None = None
        self._cancel_requested = False
        self._is_executing = False
        self._disposed = False
        self._subscriptions = [
            t.subscribe(lambda _: self.raise_can_execute_changed()) for t in triggers
        ]

    # ------------------------------------------------------------------
    # Command protocol implementation
    # ------------------------------------------------------------------

    @property
    def is_executing(self) -> bool:
        """True while an execution is in flight; False when idle."""
        with self._gate:
            return self._is_executing

    def can_execute(self, parameter: object = None) -> bool:
        """Return False while in flight; otherwise the predicate result (or True).

        A predicate that raises is treated as False (defensive).
        """
        with self._gate:
            if self._disposed or self._is_executing:
                return False
        if self._predicate is None:
            return True
        try:
            allowed = self._predicate()
        except Exception:
            return False
        with self._gate:
            return allowed and not self._disposed and not self._is_executing

    async def execute_async(self, parameter: object = None) -> None:
        """Run the async task once, observing cancellation.

        No-op when ``can_execute()`` is False or no task is configured. By default
        a cancellation requested via :meth:`cancel` is swallowed and this coroutine
        returns normally; with ``throw_on_cancel=True`` the ``CancelledError`` is
        re-raised to the awaiter.
        """
        task_factory = self._task
        if task_factory is None:
            return
        with self._gate:
            if self._disposed or self._is_executing:
                return
        if self._predicate is not None:
            try:
                if not self._predicate():
                    return
            except Exception:
                return
        with self._gate:
            if self._disposed or self._is_executing:
                return
            self._cancel_requested = False
            self._is_executing = True
        self.raise_can_execute_changed()
        inner: asyncio.Task[None] = asyncio.ensure_future(task_factory())
        with self._gate:
            self._current_task = inner
            self._current_loop = asyncio.get_running_loop()
            cancel_immediately = self._cancel_requested or self._disposed
        if cancel_immediately:
            inner.cancel()
        try:
            await inner
        except asyncio.CancelledError:
            # Non-throwing default (DIA-007 alignment): swallow only the
            # cancellation WE requested via cancel(); re-raise an external one so
            # asyncio cancellation semantics are preserved.
            with self._gate:
                cancel_requested = self._cancel_requested
            if self._throw_on_cancel or not cancel_requested:
                raise
        finally:
            with self._gate:
                self._is_executing = False
                self._current_task = None
                self._current_loop = None
            self.raise_can_execute_changed()

    def execute(self, parameter: object = None) -> None:
        """Fire-and-forget. Schedules ``execute_async`` on the current event loop.

        A faulting task is routed to the :attr:`errors` channel instead of being
        swallowed; a cancellation is already swallowed by the non-throwing default.
        """
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            future = submit_background(self.execute_async(parameter))
            future.add_done_callback(self._on_done)
            return
        task = loop.create_task(self.execute_async(parameter))
        task.add_done_callback(self._on_done)

    def cancel(self) -> None:
        """Request cancellation of the in-flight task; a no-op when idle."""
        with self._gate:
            self._cancel_requested = True
            task = self._current_task
            loop = self._current_loop
        if task is not None and not task.done():
            if loop is not None and loop.is_running():
                loop.call_soon_threadsafe(task.cancel)
            else:
                task.cancel()

    @property
    def can_execute_changed(self) -> rx.Observable[None]:
        """Observable that emits ``None`` on each trigger and in-flight state flip.

        The backing Subject is sealed behind ``as_observable`` so callers can only
        subscribe — never ``on_next``/``dispose`` the internal stream (VMX-013).
        """
        return self._can_execute_changed_subject.pipe(ops.as_observable())

    @property
    def errors(self) -> rx.Observable[BaseException]:
        """Observable surfacing a fault from the fire-and-forget :meth:`execute`.

        ``execute()`` returns immediately, so a throwing task cannot propagate to
        the caller; instead of swallowing it the error is emitted here. Await
        :meth:`execute_async` to observe it inline. Cancellations never reach this
        channel. Completes on :meth:`dispose`.
        """
        return self._errors.pipe(ops.as_observable())

    def _on_done(self, task: asyncio.Future[None] | Future[None]) -> None:
        if task.cancelled():
            return
        exc = task.exception()
        if exc is not None:
            self._emit_error(exc)

    def _emit_error(self, exc: BaseException) -> None:
        with self._gate:
            if self._disposed:
                return
        self._errors.on_next(exc)

    def raise_can_execute_changed(self) -> None:
        """Emit one re-evaluation notification without invoking user delegates.

        Valid while idle or in flight; repeated calls are additive. Calls after
        :meth:`dispose` are no-ops.
        """
        with self._gate:
            if self._disposed:
                return
        self._can_execute_changed_subject.on_next(None)

    def dispose(self) -> None:
        """Cancel any in-flight task, release subscriptions, complete the subjects.

        Idempotent: subsequent calls are a no-op.
        """
        with self._gate:
            if self._disposed:
                return
            self._disposed = True
        self.cancel()
        for sub in self._subscriptions:
            sub.dispose()
        self._can_execute_changed_subject.on_completed()
        self._can_execute_changed_subject.dispose()
        self._errors.on_completed()
        self._errors.dispose()

    # ------------------------------------------------------------------
    # Builder entry-point
    # ------------------------------------------------------------------

    @staticmethod
    def builder() -> AsyncRelayCommandBuilder:
        """Return a new immutable builder for ``AsyncRelayCommand``."""
        return AsyncRelayCommandBuilder()


@dataclasses.dataclass(frozen=True)
class AsyncRelayCommandBuilder:
    """Immutable fluent builder for :class:`AsyncRelayCommand`.

    Each setter returns a NEW builder instance via ``dataclasses.replace``
    (satisfies BLD-001).
    """

    _task: Callable[[], Awaitable[None]] | None = dataclasses.field(default=None)
    _predicate: Callable[[], bool] | None = dataclasses.field(default=None)
    _triggers: tuple[rx.Observable[object], ...] = dataclasses.field(default_factory=tuple)
    _throw_on_cancel: bool = dataclasses.field(default=False)

    def task(self, callable_: Callable[[], Awaitable[None]]) -> AsyncRelayCommandBuilder:
        """Set the cancellable async task. Returns a new builder."""
        return dataclasses.replace(self, _task=callable_)

    def predicate(self, callable_: Callable[[], bool]) -> AsyncRelayCommandBuilder:
        """Set the predicate that gates can_execute. Returns a new builder."""
        return dataclasses.replace(self, _predicate=callable_)

    def triggers(self, observable: rx.Observable[object]) -> AsyncRelayCommandBuilder:
        """Add a trigger observable (additive). Returns a new builder."""
        return dataclasses.replace(self, _triggers=(*self._triggers, observable))

    def throw_on_cancel(self, value: bool = True) -> AsyncRelayCommandBuilder:
        """Opt into re-raising the cancellation to the awaiter. Returns a new builder."""
        return dataclasses.replace(self, _throw_on_cancel=value)

    def build(self) -> AsyncRelayCommand:
        """Build and return the :class:`AsyncRelayCommand` (succeeds even with no task)."""
        return AsyncRelayCommand(
            self._task,
            self._predicate,
            list(self._triggers),
            throw_on_cancel=self._throw_on_cancel,
        )
